velox/functions/sparksql/specialforms/DecimalRound.cpp

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/specialforms/DecimalRound.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+class DecimalRoundFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    const auto& inputType = args[0]->type();
+    auto [inputPrecision, inputScale] = getDecimalPrecisionScale(*inputType);
+    auto [resPrecision, resScale] = getDecimalPrecisionScale(*resultType);
+    int32_t scale = 0;
+    if (args.size() > 1) {
+      VELOX_USER_CHECK(args[1]->isConstantEncoding());
+      scale = args[1]->asUnchecked<ConstantVector<int32_t>>()->valueAt(0);
```

Comment:
Can scale be different than resScale?

Reply:
Yes, that's true. There are some factors being considered in the result scale calculation. When `scale` is negative the result scale will be 0, and when `scale` is greater than the scale of input decimal, the minor one will be the result scale.
Below are the details in Spark.
https://github.com/apache/spark/blob/branch-3.3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala#L1430-L1444

Reply:
Add a check in the constructor to verify resultPrecision and resultScale.